### PR TITLE
Dragdrop

### PR DESCRIPTION
**Pull request by dspoerr**
Adds drag and drop functionality to assignments, and removes the drop down lists. Player names are now moveable and can be dragged to groups. Does not allow for multiple assignments anymore.

### CODE REVIEW

File: ClassicHealAssignments.lua [hunk 1]
```diff
@@ -102,31 +109,30 @@ function UpdateFrame()
...
+               nameFrame:SetCallback("OnDragStart", function(widget) DragStart(widget) end )    
+               nameFrame:SetCallback("OnDragStop", function(widget) DragStop(widget) end)
                playerFrames[player] = nameFrame
-               healerGroup:AddChild(nameFrame)
+               local nameContainer = AceGUI:Create("SimpleGroup")
+               nameContainer:SetRelativeWidth(1)
+               nameContainer:AddChild(nameFrame)
+
+			   -- add the player frame to the healer container only if they are unassigned
+			   if reverseAssignments[player] == nil then
```
Comment (csundlof) [inline ClassicHealAssignments.lua:125]: nit: fix indentation for this if statement and the surrounding lines (the if-statement should be on the same level of indentation as the code above)

File: ClassicHealAssignments.lua [hunk 2]
```diff
@@ -102,31 +109,30 @@ function UpdateFrame()
...
+			   end
             end
 
             local playerFrameText = player
-
-            if not table.isEmpty(GetAssignmentsForPlayer(player)) then
-               playerFrameText = playerFrameText .. "(X)"
-            end
-
             playerFrames[player]:SetText(playerFrameText)
```
Comment (csundlof) [inline ClassicHealAssignments.lua:131]: You can set this text at the creation of the player frame now instead. So move this up, and just do `nameframe:SetText(player)` instead, and delete the playerFrameText variable.

File: ClassicHealAssignments.lua [hunk 3]
```diff
@@ -102,31 +109,30 @@ function UpdateFrame()
...
-
             tinsert(healerList, player)
-            tinsert(dispellerList, player)
```
Comment (csundlof) [inline ClassicHealAssignments.lua:122]: the dispellerList variable isn't used anymore right? you can delete it from the file

File: ClassicHealAssignments.lua [hunk 4]
```diff
@@ -471,3 +449,101 @@ function GetAssignmentsForPlayer(player)
       return {}
    end
 end
+
+
+function DragStart(widget)
+   widget.frame:ClearAllPoints()
+   widget.frame:StartMoving()
+   local cursorX, cursorY = GetCursorPosition()
```
Comment (csundlof) [inline ClassicHealAssignments.lua:457]: Do you need this line?

File: ClassicHealAssignments.lua [hunk 5]
```diff
@@ -471,3 +449,101 @@ function GetAssignmentsForPlayer(player)
...
+function DragStop(widget)
+   local uiScale = UIParent:GetEffectiveScale()
+   local cursorX, cursorY = GetCursorPosition()
+   local scaleCursorX = cursorX / uiScale
+   local scaleCursorY = cursorY / uiScale
+
+   local playerName = widget:GetUserData("playerName")
+
+   -- check the unassigned group
+   if scaleCursorX > healerGroup.frame:GetLeft() and scaleCursorX < healerGroup.frame:GetRight() then
```
Comment (csundlof) [inline ClassicHealAssignments.lua:470]: Should the comparisons here be >=, <= instead?

File: ClassicHealAssignments.lua [hunk 6]
```diff
@@ -471,3 +449,101 @@ function GetAssignmentsForPlayer(player)
...
+		if scaleCursorY > healerGroup.frame:GetBottom() and scaleCursorY < healerGroup.frame:GetTop() then
+		-- Cursor in the unassigned group. Just need to reset assignments
+			ClearAssignments(playerName)
+		end
+	else
+	   for assignment, frame in pairs(assignmentList) do
+		  -- check if the cursor drop is within the frame area
+		  if scaleCursorX > frame.frame:GetLeft() and scaleCursorX < frame.frame:GetRight() then
+			 if scaleCursorY > frame.frame:GetBottom() and scaleCursorY < frame.frame:GetTop() then
+
```
Comment (csundlof) [inline ClassicHealAssignments.lua:480]: nit: remove newline

File: ClassicHealAssignments.lua [hunk 7]
```diff
@@ -471,3 +449,101 @@ function GetAssignmentsForPlayer(player)
...
+	else
+	   for assignment, frame in pairs(assignmentList) do
+		  -- check if the cursor drop is within the frame area
+		  if scaleCursorX > frame.frame:GetLeft() and scaleCursorX < frame.frame:GetRight() then
+			 if scaleCursorY > frame.frame:GetBottom() and scaleCursorY < frame.frame:GetTop() then
+				
+				-- correct frame found, clearing all assignments for correct reassignment
+				ClearAssignments(playerName)
+
+				-- set assignments, initialize if the tables are empty
```
Comment (csundlof) [inline ClassicHealAssignments.lua:484]: Break the assignment logic here out into another function. Look at the old AssignHealer function for inspiration.

File: ClassicHealAssignments.lua [hunk 8]
```diff
@@ -471,3 +449,101 @@ function GetAssignmentsForPlayer(player)
...
+				if(assignedHealers[assignment] ~= nil) then
+				   tinsert(assignedHealers[assignment],playerName)
```
Comment (csundlof) [inline ClassicHealAssignments.lua:486]: n
```suggestion
				   tinsert(assignedHealers[assignment], playerName)
```

File: ClassicHealAssignments.toc [hunk 9]
```diff
@@ -9,4 +9,10 @@ ClassicHealAssignments.lua
...
 utils\helpers.lua
-ui\customlayouts.lua
\ No newline at end of file
+<<<<<<< HEAD
+ui\customlayouts.lua
+ui\AceGUIWidget-DragLabel.lua
+ui\AceGUIContainer-InlineGroupMod.lua
+=======
+ui\customlayouts.lua
+>>>>>>> csundlof/master
```
Comment (csundlof) [inline ClassicHealAssignments.toc:18]: bad merge here (just fix it manually and commit the fix)

File: ui/AceGUIWidget-DragLabel.lua [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	frame:SetScript("OnDragStart", Control_OnDragStart)
+   frame:SetScript("OnDragStop", Control_OnDragStop)
+
+   for method, func in pairs(methods) do
+		widget[method] = func
+	end
+
+	return AceGUI:RegisterAsWidget(widget)
+
+--	local frame = label.frame
```
Comment (csundlof) [inline ui/AceGUIWidget-DragLabel.lua:167]: delete this commented out code

File: ClassicHealAssignments.lua (hunk 1 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:125]: fixed indentation for this section

File: ClassicHealAssignments.lua (hunk 2 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:131]: moved to line 120, immediately after the frame is added to the playerFrames array.

File: ClassicHealAssignments.lua (hunk 3 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:122]: Removed all instances of dispellerList. Noticed that playerList was not a needed arg of the current version of CreateAssignmentGroup. Cleaned up references to that.

File: ClassicHealAssignments.lua (hunk 4 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:457]: not needed. cursorX and cursorY are used in the DragStop function

File: ClassicHealAssignments.lua (hunk 5 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:470]: fixed

File: ClassicHealAssignments.lua (hunk 6 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:480]: removed newline at 480

File: ClassicHealAssignments.lua (hunk 7 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:484]: added new function AssignHealer

File: ClassicHealAssignments.lua (hunk 8 above)
Comment (dspoerr) [reply ClassicHealAssignments.lua:486]: added space

File: ClassicHealAssignments.toc (hunk 9 above)
Comment (dspoerr) [reply ClassicHealAssignments.toc:18]: fixed

File: ui/AceGUIWidget-DragLabel.lua (hunk 10 above)
Comment (dspoerr) [reply ui/AceGUIWidget-DragLabel.lua]: cleaned

File: ClassicHealAssignments.lua [hunk 11]
```diff
@@ -525,25 +508,42 @@ end
...
+            if table.isEmpty(assignedHealers[assignment]) then
+               assignedHealers[assignment] = nil
+            end
+         end
+      end
+   end
 end
+
+function AssignHealer(assignment, playerName)
+
```
Comment (csundlof) [inline ClassicHealAssignments.lua:535]: nit: remove newline here

Comment (dspoerr) [reply ClassicHealAssignments.lua]: done

**Pull request merged**